packages/flutter_tools/lib/src/dart/pub.dart

```diff
@@ -35,6 +35,51 @@ const int _kPubExitCodeUnavailable = 69;
 
 typedef MessageFilter = String? Function(String message);
 
+/// globalCachePath is the directory in which the content of the localCachePath will be moved in
+void joinCaches({
```

Comment:
This has the potential to create broken `PUB_CACHE` folders.
If an error or crash occurs in the middle of these copying operations the user will be left with an invalid `PUB_CACHE`.

Running `flutter pub cache repair` should fix it. But there is no good mechanism to communicate this to the user.

In pub we do extract to a temporary folder and then do atomic move operations. This has downsides that Windows AV sometimes interferes and makes the move operation fail because it's locking files. So moving happens in a retry loop with a large number of retries :see_no_evil: 

It's entirely possible that copying in-place and hoping for the best is a viable option.

Reply:
That is a really good point I see a lot of potential issues filled if this breaks I guess we should have a fallback if something goes wrong and keep the local pub cache and next time this gets called it will try to do the whole process again. What do you think? @christopherfujino

Reply:
After doing a test that consisted in copying everything inside the `global-cache` into a `temp-directory` then doing the `joinCaches` with `global-directory` and `local-cache` it took an extra **50038 ms** (for 126 dependencies)

Current implementation: 13348 ms
Using a temp directory: 63386 ms

So I believe the current implementation should stay with the addition of having a try except on each copy and if at some point it fails delete that directory/dependency (on the global cache) and keep both caches (global and local) until the next time the user hits this part of the code and retry the join of the caches.

Reply:
SGTM